lib/improver/cli/__init__.py

```diff
@@ -0,0 +1,416 @@
+# POSSIBILITY OF SUCH DAMAGE.
+"""init for cli and clize"""
+
+from collections import OrderedDict
+
+import clize
+from clize import parameters
+from clize.help import (
+    HelpForAutodetectedDocstring,
+    ClizeHelp)
+from clize.parser import value_converter
+from clize.runner import Clize
+from sigtools.wrappers import decorator
+
+# selected clize imports/constants
+
+IGNORE = clize.Parameter.IGNORE
+LAST_OPTION = clize.Parameter.LAST_OPTION
+REQUIRED = clize.Parameter.REQUIRED
+UNDOCUMENTED = clize.Parameter.UNDOCUMENTED
+
+
+# help helpers
+
+
+def docutilize(obj):
+    """Convert Numpy or Google style docstring into reStructuredText format.
+
+    Args:
+        obj (str or obj):
+            Takes an object and changes it's docstrings to a reStructuredText
+            format.
+    Returns:
+        (str or obj):
+            A converted string or an object with replaced docstring depending
+            on the type of the input.
+    """
+    from inspect import cleandoc, getdoc
+    from sphinx.ext.napoleon.docstring import GoogleDocstring, NumpyDocstring
+    if isinstance(obj, str):
+        doc = cleandoc(obj)
+    else:
+        doc = getdoc(obj)
+    doc = str(NumpyDocstring(doc))
+    doc = str(GoogleDocstring(doc))
+    if isinstance(obj, str):
+        return doc
+    obj.__doc__ = doc
+    return obj
+
+
+class HelpForNapoleonDocstring(HelpForAutodetectedDocstring):
+    """Help for Napolean Docstrings."""
+    def add_docstring(self, docstring, *args, **kwargs):
+        """Adds the updated docstring."""
+        docstring = docutilize(docstring)
+        super().add_docstring(docstring, *args, **kwargs)
+
+
+class DocutilizeClizeHelp(ClizeHelp):
+    """Subclass to build Napoleon docstring from subject."""
+    def __init__(self, subject, owner,
+                 builder=HelpForNapoleonDocstring.from_subject):
+        super().__init__(subject, owner, builder)
+
+
+# input handling
+
+
+class ObjectAsStr(str):
+    """Hide object under a string to pass it through Clize parser."""
+    __slots__ = ('original_object',)
+
+    def __new__(cls, obj, name=None):
+        if isinstance(obj, cls):  # pass object through if already wrapped
+            return obj
+        if name is None:
+            name = cls.obj_to_name(obj)
+        self = str.__new__(cls, name)
+        self.original_object = obj
+        return self
+
+    @staticmethod
+    def obj_to_name(obj, cls=None):
+        """Helper function to create the string."""
+        if cls is None:
+            cls = type(obj)
+        try:
+            obj_id = hash(obj)
+        except TypeError:
+            obj_id = id(obj)
+        return '<%s.%s@%i>' % (cls.__module__, cls.__name__, obj_id)
+
+
+def maybe_coerce_with(convert, obj, **kwargs):
+    """Apply converter if str, pass through otherwise."""
+    obj = getattr(obj, 'original_object', obj)
+    return convert(obj, **kwargs) if isinstance(obj, str) else obj
+
+
+@value_converter
+def input_cube(to_convert):
```

Comment:
Note that the names of converter functions will end up in help and usage messages as uppercase description of arguments, so we should name them with that in mind. I personally prefer to see a single word, e.g, `--mask=INPUTCUBE` rather than snake case `--mask=INPUT_CUBE`. That's at least what most of the CLIs I've seen are doing (e.g., git will use `refspec` instead of `ref_spec`), but I'm sure opinions will vary.